### ISSUE TITLE
"uncaught-error handler" called too early

### ISSUE THREAD

**Issue by nielsenko** [dart-lang/language#2502]
The following program does not print `caught Exception: canceled`, as one might expect:
```dart
import 'dart:async';

Future<void> nothing() async {}

Future<void> main() async {
  try {
    final f = Future.error(Exception("canceled"));
    await nothing();
    await f; // we do attach a handler, but too late
  } catch (e) {
    print('caught $e'); // <-- not hit
  }
}
```
I think that is counter-intuitive.

**Comment by Levi-Lesches** [comment]
According tot he docs for [`Future.error`](https://api.dart.dev/stable/2.18.1/dart-async/Future/Future.error.html), it creates a microtask which completes with the error. Microtasks finish before the next event, meaning before the next `await` is completed. You can see this in the following snippet as well: 
```dart
import 'dart:async';

Future<void> nothing() async {}

Future<void> main() async {
  scheduleMicrotask(() => print("Microtask completed"));
  await nothing();
  print("Done awaiting f");
}

```

**Comment by lrhn** [comment]
The goal of the current design is to never let an error be lost by accident.
If a future completes with an error, and error will be reported, either to the listeners, or to the global/zone uncaught error handler.

The problem with that goal is that we can't just allow a handler to be attached at "some later point", because that might turn out to be never. We do not want to retain all error futures forever and spew all the errors out when the program exits. If the program exits in a controlled way. Web programs don't exit. Native programs which call `exit` don't leave time to do anything.

So, we need to decide a maximum time we want to allow waiting from an error future is completed until it must be handled.
We chose zero time, the only non-arbitrary choice. We also made sure at least one microtask passes between asking for a future to complete with an error, and it actually completing, so you always have *some* time to add an error handleer.

In other words, the current behavior is deliberate and considered the best tradeoff that achieves the original goal.
